### PR TITLE
LIME-201: Add user card

### LINKED ISSUE #201: Add User card
Image

### PR DESCRIPTION
**Pull request by DavidCharif**
image

### CODE REVIEW

**Comment by DavidCharif** [comment]
https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-lime/assets/53876827/cc26863e-50d9-4823-81d1-452b4467ea2f

**Comment by stpalamar** [comment]
I think better use icon from figma design. I know that we use heroicons but its message icons looks bad in our design. Or maybe try to make it smaller

**Comment by DavidCharif** [comment]
image
image
image

File: frontend/src/bundles/friends/components/friend-card/friend-card.tsx [hunk 1]
```diff
@@ -0,0 +1,110 @@
+import { useCallback } from 'react';
```
Comment (anton-otroshchenko) [inline frontend/src/bundles/friends/components/friend-card/friend-card.tsx]: import from hooks.ts file

File: frontend/src/bundles/friends/components/friend-card/friend-card.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    isFriend: boolean;
+    addFriend: (id: number) => void;
+    messageFriend: (id: number) => void;
+};
+
+const isValidImg = (url: string): boolean => {
+    const img = new Image();
+    img.src = url;
+    return img.complete;
+};
```
Comment (anton-otroshchenko) [inline frontend/src/bundles/friends/components/friend-card/friend-card.tsx]: i think its better to move this into helpers folder

**Review by anton-otroshchenko — CHANGES_REQUESTED** [review]

File: frontend/src/bundles/common/helpers/check-url-img-is-valid/check-url-img-valid.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+const isValidUrlImg = (url: string): boolean => {
```
Comment (dimapopovych) [inline frontend/src/bundles/common/helpers/check-url-img-is-valid/check-url-img-valid.ts]: name should start with verb. `validateImageUrl` would be nice, also folder should have the same name as helper itself

**Review by dimapopovych — COMMENTED** [review]
also I think we don't need svg for this dot

File: frontend/src/bundles/common/helpers/check-url-img-is-valid/check-url-img-valid.ts (hunk 3 above)
Comment (DavidCharif) [reply frontend/src/bundles/common/helpers/check-url-img-is-valid/check-url-img-valid.ts]: image
As for the state dot, what do you suggest?

Comment (stpalamar) [reply frontend/src/bundles/common/helpers/check-url-img-is-valid/check-url-img-valid.ts]: You can create dot with tailwindcss) Simply `rounded-[50%]`
Like this https://play.tailwindcss.com/cRcwJASqdW

**Comment by DavidCharif** [comment]
image
image

**Review by dimapopovych — DISMISSED** [review]

File: frontend/src/bundles/friends/components/friend-card/friend-card.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+            <div className="h-3/4 w-full">
+                {validateImageUrl(avatar) ? (
+                    <img
+                        src={avatar}
+                        alt={name}
+                        className="aspect-square rounded-t-xl object-cover"
+                    />
+                ) : (
+                    <div className="bg-lm-grey-100 flex aspect-square items-center justify-center rounded-t-xl">
+                        <p>Error loading image</p>
```
Comment (nikita-remeslov) [inline frontend/src/bundles/friends/components/friend-card/friend-card.tsx]: use some placeholder image

File: frontend/src/assets/img/icons/message-icon.svg
Comment (nikita-remeslov) [inline frontend/src/assets/img/icons/message-icon.svg:1]: https://heroicons.com/

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]
show light theme example

File: frontend/src/assets/img/icons/message-icon.svg
Comment (DavidCharif) [reply frontend/src/assets/img/icons/message-icon.svg:1]: https://github.com/BinaryStudioAcademy/bsa-winter-2023-2024-lime/pull/209#issuecomment-1967151634

This is closest similar to the figma svg from heroicons, 
image
And this is the SVG from figma 
image

**Comment by DavidCharif** [comment]
image
image
image
image

File: frontend/src/bundles/common/helpers/validate-image-url/validate-image-url.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+const validateImageUrl = (url: string): boolean => {
```
Comment (dimapopovych) [inline frontend/src/bundles/common/helpers/validate-image-url/validate-image-url.ts]: file should be named
validate-image-url.helper.ts

File: frontend/src/bundles/friends/components/friend-card/friend-card.tsx [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    const variantButton = isFriend ? 'secondary' : 'primary';
+    const classesButtonResponsive =
+        'lg:px-4 lg:py-2 lg:h-8  sm:text-[0.7rem] sm:h-6 sm:px-1 sm:py-1';
+    const labelsButton = {
+        add: 'Add friend',
+        remove: 'Remove friend',
+    };
+    const icon = 'messageIcon';
+    const size = 'sm';
```
Comment (dimapopovych) [inline frontend/src/bundles/friends/components/friend-card/friend-card.tsx]: use values from enums

File: frontend/src/bundles/friends/components/friend-card/friend-card.tsx [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+        messageFriend(id);
+    }, [messageFriend, id]);
+
+    const variantButton = isFriend ? 'secondary' : 'primary';
+    const classesButtonResponsive =
+        'lg:px-4 lg:py-2 lg:h-8  sm:text-[0.7rem] sm:h-6 sm:px-1 sm:py-1';
+    const labelsButton = {
+        add: 'Add friend',
+        remove: 'Remove friend',
+    };
```
Comment (dimapopovych) [inline frontend/src/bundles/friends/components/friend-card/friend-card.tsx]: I think all of these could be inlined

**Review by dimapopovych — CHANGES_REQUESTED** [review]

**Review by stpalamar — DISMISSED** [review]

File: frontend/src/bundles/friends/enums/icon-friend-card.enums.ts [hunk 8]
```diff
@@ -0,0 +1,12 @@
...
+import { ComponentSize } from '~/bundles/common/enums/component-size.enum.js';
+
+const IconFriendCard = {
+    ICON: IconName.messageIcon,
+    SIZE: ComponentSize.MEDIUM,
+    COLOR_IS_FRIEND: IconColor.PRIMARY,
+    COLOR_NOT_FRIEND: IconColor.SECONDARY,
+};
+
+export { IconFriendCard };
```
Comment (dimapopovych) [inline frontend/src/bundles/friends/enums/icon-friend-card.enums.ts]: seems unnecessary for me

**Review by dimapopovych — CHANGES_REQUESTED** [review]

**Review by ArtemLymarenko — DISMISSED** [review]

**Review by stpalamar — DISMISSED** [review]

**Comment by DavidCharif** [comment]
image
image
image
image

File: frontend/src/bundles/friends/components/friend-card/friend-card.tsx [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+import { validateImageUrl } from '~/bundles/common/helpers/helpers.js';
+import { useCallback } from '~/bundles/common/hooks/hooks.js';
+
+type FriendProperties = {
+    id: number;
+    name: string;
+    avatarUrl: string;
+    isActive: boolean;
+    isFollow: boolean;
+    toggleFollow: (id: number, isFollow: boolean) => void;
```
Comment (dimapopovych) [inline frontend/src/bundles/friends/components/friend-card/friend-card.tsx]: should be isFollowed and onFollogToggle or something, by quality criteria

**Review by dimapopovych — CHANGES_REQUESTED** [review]

**Comment by DavidCharif** [comment]
image
image

**Review by nikita-remeslov — APPROVED** [review]

**Pull request merged**